atsign-foundation/at_client_sdk
# feat: client commit log compaction

**- What I did**
* Introduce client-side commit log compaction

**- How I did it**
* Introduced `AtClientCommitLogCompaction` in at_client package that calls the compaction job in the at_persistence_secondary_server package.
* Introduced a new variable "commitLogCompactionTimeIntervalInMins" in AtClientConfig that sets the time interval of Compaction Job.
* Changed AtClientImpl as a class implementing "AtSignChangeListener". On switch atSign, the compaction cron job of the previous atSign is stopped and the cron job on the new atSign is started.
* In "listenToAtSignChange" method, added the IF condition to check and remove only the previous atSign AtClientImpl instance.
   - Reintroduced the IF condition because the AtClientImpl instance is created before the notifyListeners method.
   - notifyListeners method has switchAtSignEvent as an argument and SwitchAtSignEvent needs CurrentAtClient Instance. So did not move notifyListeners method before the creation of AtClientImpl.
* On switch atSign event, removing the atSign from the static map instance - "atClientInstanceMap" for the following reasons
   * On switch event, to avoid returning the old AtClientImpl instance which is removed from AtClientManager._changeListeners.
      - If the same instance is returned, since it is not in AtClientManager._changeListener upon subsequent switch atSign, "this" AtClientImpl instance will not be notified.
* Added getter on "AtClientCommitLogCompaction" to expose "isCompactionJobRunning" method to assert if the compaction job is running in e2e tests.
* In at_client_service.dart, removed the usage of SyncService from AtClientManger since it is deprecated and replaced it with SyncService from AtClientImpl.
      
**- How to verify it**

* Updated the unit tests in at_client_impl_test.dart to assert the previous atSign compaction job listener is removed on switch atSign events
* Update the unit tests in at_persistence_secondary_server with tests that mimic client-side storage.
  - In the PR: https://github.com/atsign-foundation/at_server/pull/950 , packages/at_persistence_secondary_server/test/at_compaction_test.dart contains the unit tests.
* Added a functional test to verify the commit log compaction on a single atSign. Call the "put" request on the same key to create multiple commit entries and assert that entries in the commit log shrink after the compaction.
* Added end2end test to asset the commit log compaction on a switch atSign scenario and a test to verify that commit log shrinks on compaction 


**- Description for the changelog**
* feat: Introduce client-side commit log compaction.

Additional Notes:

* Set the dependency overrides to at_persistence_secondary_server to expose AtCompactionService to skip the cron job and run compaction job for testing purpose. Following is the PR : https://github.com/atsign-foundation/at_server/pull/1156
* @gkc: Since "AtCompactionService.executeCompaction" has an input argument of "AtLogType", exported the "AtCompactionService" class in at_persistence_secondary_server. The tests would directly invoke "AtCompactionService.executeCompaction" by passing the client-side commit log to compact.
* Skipped the changes in "AtCompactionJob.scheduleCompactionJob" to skip the cron job since the above change solves the issue.
* Introduced "isScheduled" method in "AtCompactionJob" class which indicates if the compaction job is running or not.
